storage2/src/storage.rs

```diff
@@ -29,16 +64,129 @@ impl Storage {
     }
 
     // TODO: this probably can't be 'static long-term
-    pub async fn apply(&'static mut self, state: State) {
-        // TODO: 1. write the index tables and JMT to RocksDB
-        // 2. update the snapshot
-        // TODO: set jmt_version correctly
-        let jmt_version = 0;
-        let snapshot = self.db.snapshot();
-        // Obtain the write-lock for the latest snapshot, and replace it with the new snapshot.
-        let mut guard = self.latest_snapshot.write();
-        *guard = Arc::new(Snapshot::new(snapshot, jmt_version));
-        // Drop the write-lock (this will happen implicitly anyways, but it's good to be explicit).
-        drop(guard);
+    pub async fn apply(&'static mut self, state: State, nct: &Tree) -> Result<()> {
+        // 1. Write the NCT
+        tracing::debug!("serializing NCT");
+        let tct_data = bincode::serialize(nct)?;
+        tracing::debug!(tct_bytes = tct_data.len(), "serialized NCT");
+        gauge!(metrics::NCT_SIZE_BYTES, tct_data.len() as f64);
+
+        let db = self.db;
+
+        let span = Span::current();
+        tokio::task::Builder::new()
+            .name("put_nct")
+            .spawn_blocking(move || {
+                span.in_scope(|| {
+                    let nct_cf = db.cf_handle("nct").expect("nct column family not found");
+                    db.put_cf(nct_cf, "tct", &tct_data)
+                })
+            })
+            .unwrap()
+            .await??;
+
+        // 2. Write the JMT to RocksDB
+        // We use wrapping_add here so that we can write `new_version = 0` by
+        // overflowing `PRE_GENESIS_VERSION`.
+        let old_version = self.latest_version().await?.unwrap();
+        let new_version = old_version.wrapping_add(1);
+        tracing::trace!(old_version, new_version);
+        let span = Span::current();
+        tokio::task::Builder::new()
+            .name("Storage::write_node_batch")
+            .spawn_blocking(move || {
+                span.in_scope(|| {
+                    let snap = self.latest_snapshot.read().clone();
+                    let jmt = JellyfishMerkleTree::new(snap.as_ref());
+                    // Write the unwritten changes from the state to the JMT.
+                    let (jmt_root_hash, batch) = jmt.put_value_set(
+                        state
+                            .unwritten_changes
+                            .into_iter()
+                            .map(|x| (KeyHash::from(x.0), x.1))
+                            .collect(),
```

Comment:
This `.collect()` is now unnecessary (and I think will lead to a type ambiguity error) due to the change I made to the `put_value_set` function: now it takes an `IntoIterator` (and since `Iterator`s implement `IntoIterator`, this will Just Work:tm:).

Reply:
This isn't working right for some reason:

```rust
mismatched types
expected struct `std::vec::Vec<(jmt::KeyHash, std::option::Option<std::vec::Vec<u8>>)>`
   found struct `std::iter::Map<std::collections::btree_map::IntoIter<std::string::String, std::option::Option<std::vec::Vec<u8>>>, [closure@storage2/src/storage.rs:106:34: 106:63]>`
```

Reply:
despite rust-analyzer saying that it takes an `IntoIterator`... maybe something's cached on my local

Reply:
Try doing `cargo update` and see if that fixes it.

Reply:
I did, it's weirdly only happening in VSCode, on the command-line when I `cargo build` it's fine

Reply:
Have you restarted rust-analyzer since you did the cargo update?

Reply:
Yeah, I wound up working around it by manually specifying the revision in the `Cargo.toml` 🙃

Reply:
I don't think that should be necessary. I would try `cargo clean && cargo update`. Something is cached that shouldn't be.